Add lambda-otel4s module

alexcardell opened the pull request:
Supercedes #456 

- Adds lambda-natchez module
- Adds lambda-otel4s module
  - Traces lambdas with semantic conventions applied
- Adds otel4s examples
- Add scalafix migration to port natchez package imports -- (removes previous version scalafix migrations)

This is targeting otel4s **1.0.0-RC1** ~0.16.x~ ~0.15.x~ ~0.14.x~ ~0.13.x~ ~0.12.x~ ~0.11.x~ ~0.10.x~ ~0.9.x~ ~0.8.x~ ~0.7.x~ ~0.6.x~ ~0.5.x~

# Links

https://opentelemetry.io/docs/specs/semconv/faas/aws-lambda/
https://opentelemetry.io/docs/specs/semconv/faas/faas-spans/

alexcardell committed:
Extract lambda-natchez module

This is to allow consumers the choice of natchez
or otel4s

No breaking changes except requiring a
`feral.lambda.natchez._` import

alexcardell committed:
Remove natchez-core dependency from lambda module

alexcardell committed:
Add lambda-otel4s module

alexcardell committed:
Use SqsEvent for attributes, not SqsRecord

alexcardell committed:
Add otel4s-sdk example

alexcardell committed:
re-run prePR

alexcardell committed:
Remove resolvers

alexcardell committed:
Add otel4s TracedHandler tests

alexcardell committed:
Rework package structure

alexcardell committed:
prePR

alexcardell commented:
Currently pointed at a local snapshot so CI will definitely fail, as otel4s main has `NoPublish` for everything involving the SDK

alexcardell committed:
Swap tests from SDK to oteljava

alexcardell committed:
Run prePR

alexcardell committed:
Add back extra JVM test

alexcardell committed:
Add OtelJava JVM example

iRevive commented on lambda-otel4s/shared/src/main/scala/feral/lambda/otel4s/FaasAttributes.scala:
```diff
@@ -0,0 +1,47 @@
+import org.typelevel.otel4s.semconv.resource.attributes.ResourceAttributes
+import org.typelevel.otel4s.semconv.trace.attributes.SemanticAttributes
+
+object LambdaContextTraceAttributes {
```
Since semantic conventions 1.24.0, attributes are split into stable and experimental.

Stable attributes are available in the `otel4s-semconv` package and experimental in the `otel4s-semconv-experimental`. 

Also, attributes aggregated by the namespaces, for example, [FaasExperimentalAttributes](https://github.com/typelevel/otel4s/blob/main/semconv/experimental/src/main/scala/org/typelevel/otel4s/semconv/experimental/FaasExperimentalAttributes.scala).

Experimental attributes may have binary incompatible changes between releases, so libraries should not use them directly. 

____

Can feral pull the `otel4s-semconv-experimental`?

alexcardell replied:
Whether libraries should pull that in is an interesting question, but it appears all of the FAAS/Cloud attributes are in experimental, so I've added it. Otherwise there aren't really any attributes to add

alexcardell replied:
Options:
1. Add `experimental`
2. Remove experimental attributes and wait for them to be stable
3. Re-use attribute keys to create new attributes inline, and port to experimental later

iRevive commented on lambda-otel4s/shared/src/main/scala/feral/lambda/otel4s/TracedHandler.scala:
```diff
@@ -0,0 +1,55 @@
+      context <- inv.context
+      res <- Tracer[F].joinOrRoot(attr.contextCarrier(event)) {
+        buildSpan(event, context).surround {
+          for {
+            res <- handler(event)
+          } yield res
```
```suggestion
          handler(event)
```

iRevive commented on lambda-otel4s/shared/src/main/scala/feral/lambda/otel4s/package.scala:
```diff
@@ -0,0 +1,34 @@
+ */
+
+package feral.lambda
+/*
+ * Copyright 2021 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
Headers got funky

alexcardell replied:
Fixed in 640fabf

alexcardell committed:
Update otel4s to 0.6.0

alexcardell committed:
Simplify for-comp

Co-authored-by: Maksym Ochenashko <maxochenashko@gmail.com>

alexcardell committed:
Fix headers

alexcardell committed:
Remove semconv-experimental

alexcardell committed:
Add attributes for other events

alexcardell committed:
Remove unused imports

alexcardell commented:
@iRevive Updated all the attributes and removed semconv-experimental

Would appreciate another look when you have time

iRevive commented:
@alexcardell the changes look good. 

We still need to wait for a release of https://github.com/http4s/http4s-otel4s-middleware, right?

alexcardell commented:
The middleware would be useful for the example, but I don't believe it's necessary otherwise

alexcardell committed:
Add http4s-otel4s-middleware to example

alexcardell committed:
Fix import order

alexcardell commented:
Upgraded to otel4s 0.7 but thanks to sbt 1.10 that required a few other updates

Also added http4s-otel4s-middleware to the example

iRevive committed:
use otel4s-sdk-trace-testkit to run tests across all platforms

iRevive commented on build.sbt:
```diff
@@ -170,20 +174,54 @@ lazy val lambdaCloudFormationCustomResource = crossProject(JSPlatform, JVMPlatfo
+    )
+  )
+  .settings(commonSettings)
+  .jvmSettings(libraryDependencies ++= Seq(
+    "org.typelevel" %%% "otel4s-oteljava-trace-testkit" % otel4sVersion % Test
+  ))
```
We can use `otel4s-sdk-trace-testkit` to run tests across all platforms. Here is an example: https://github.com/alexcardell/feral/pull/2

armanbilge commented:
I'd like to merge this for an upcoming release! @alexcardell do you have more time to work on it? @iRevive any changes in otel4s since this PR was opened that we should account for? Thanks both for your work on this, sorry it took me so long to circle back 🙏

iRevive commented:
@armanbilge I believe bumping otel4s version from `0.7.0` to `0.9.0` should be smooth.

One suggestion regarding this PR: we can switch to `otel4s-sdk-trace-testkit` to run tests across all platforms.

alexcardell commented:
@armanbilge I will fix it up!

alexcardell commented:
@iRevive thanks for the PR to my branch, sorry I did not notice that for many moons

alexcardell committed:
Move TracedHandlerSuite to JVM

alexcardell committed:
Update workflows

alexcardell committed:
Swap module order

alexcardell committed:
Fix TracedHandlerSuite class name

alexcardell committed:
Add JS TracedHandler suite

alexcardell committed:
Remove unused imports

alexcardell committed:
Update base version

alexcardell committed:
Update otel4s to 0.8

alexcardell commented:
Added JS tests for the otel4s module -- they can't be in `shared` because of the different platform Contexts, without a fair bit of work

otel4s 0.8 looks like it works fine, however I can't update to 0.9 yet until http4s/http4s-otel4s-middleware#123 is in, as it's used in the examples module

iRevive committed:
Lambda otel4s: test span attributes

iRevive commented:
One more minor improvement: https://github.com/alexcardell/feral/pull/3.
We can use `otel4s-semconv-experimental` in tests to ensure our attributes match the specification.

iRevive commented on lambda-otel4s/shared/src/main/scala/feral/lambda/otel4s/Attributes.scala:
```diff
@@ -0,0 +1,84 @@
+/**
+ * Temporary aliases for Lambda message-specific attributes in otel4s-semconv-experimental
+ */
+object LambdaMessageAttributes {
```
Should attributes be package private?

alexcardell replied:
I think this was a shim for otel4s-semconv-experimental anyway, back when I thought this would merge before that was available, I can probably replace this

iRevive replied:
`lambda-otel4s` still cannot depend on the `otel4s-semconv-experimental`, because semconv-experimental may receive binary breaking updates. And this will affect feral compatibility. 

So far, I came up with the following strategy:
1) Redefine required attributes in the library (like you did), ideally in a private package
2) Use `otel4s-semconv-experimental` in tests to ensure your attributes match the specification (even the experimental one). 
3) Once attributes get promoted to the `otel4s-semconv` stable artifact - replace handcrafted attributes with the stable one

alexcardell replied:
Added some specific attribute tests, a little bit boilerplatey but did catch a small issue

iRevive commented on lambda-otel4s/shared/src/main/scala/feral/lambda/otel4s/EventSpanAttributes.scala:
```diff
@@ -0,0 +1,38 @@
+protected trait EventSpanAttributes[E] {
+  def contextCarrier(e: E): Map[String, String]
+  def spanKind: SpanKind
+  def attributes(e: E): List[Attribute[_]]
```
As an alternative, we can use `Attributes` collection there instead.

alexcardell replied:
done

alexcardell committed:
Format

iRevive commented:
> I am not entirely clear how [aws-otel](https://aws-otel.github.io/) Lambda layers will play with this, and whether that's in scope

This one is complicated. If I get it right, it's only a predefined otel-compatible collector. Hence, it should work with otel4s-sdk and otel4s-oteljava out of the box.


Apart from the distro, there is a 'otel lambda layer': https://github.com/open-telemetry/opentelemetry-lambda.
Each language layer has a handcrafted entry point. The Java layer, for example, won't work with otel4-sdk but should work with otel4s-oteljava.

alexcardell committed:
Swap List[Attribute[_]] to Attributes

alexcardell committed:
Remove unused import

alexcardell committed:
Remove unused import

alexcardell committed:
Add LambdaContextAttributes test

alexcardell committed:
Add LambdaMessageAttributesTest

alexcardell commented:
Good to go when I can bump to otel4s 0.9, but given the middleware is only used in the examples, perhaps we can just indicate _where you would_ wrap the client with the middleware, rather than adding that dependency

iRevive commented:
There are also some AWS goodies in the upcoming `0.10.0` otel4s release: https://typelevel.org/otel4s/sdk/aws-xray-propagator.html#aws-lambda.

It depends on how urgently we need to release a new version of feral, but I'm fine with keeping examples out of the scope to unblock the release.

alexcardell committed:
Deduplicate examples module settings keys

alexcardell committed:
Make otel4s attribute utilities package private

alexcardell committed:
Remove _root_ from example imports

alexcardell committed:
Fix example

alexcardell committed:
Remove http4s-otel4s-middleware

alexcardell committed:
Bump otel to 0.9

alexcardell committed:
Fix 0.9 bump

alexcardell commented:
Fixed the 0.9 issues, good to go

alexcardell commented:
@armanbilge I'm not a bin-compat expert so maybe have a scan and see if the package private-ing will cause any issues of the like

armanbilge commented:
> There are also some AWS goodies in the upcoming `0.10.0` otel4s release (for otel4s-sdk):

@iRevive is v0.10.0 coming soon?

armanbilge commented on lambda-natchez/shared/src/main/scala/feral/lambda/natchez/package.scala:
```diff
@@ -0,0 +1,54 @@
+import feral.lambda.natchez.KernelSource
+import org.typelevel.ci._
+
+protected trait KernelSources {
```
Instead of putting these in the package object, we can put them in the companion object for `KernelSource`. Then they will automatically be in implicit scope without an import.

armanbilge replied:
@alexcardell Can you look into this as well? 🙏

alexcardell replied:
Will do

alexcardell replied:
@armanbilge done :)

iRevive commented:
> > There are also some AWS goodies in the upcoming `0.10.0` otel4s release (for otel4s-sdk):
> 
> @iRevive is v0.10.0 coming soon?

We are two PRs away from 0.10.0:
- https://github.com/typelevel/otel4s/pull/787
- https://github.com/typelevel/otel4s/pull/780

Both are implemented, some minor changes (docs, etc) may be required.

I think we will have 0.10.0 next week.

armanbilge commented:
I am going hang on for otel4s v0.10.0 to avoid back-to-back breaking releases!

iRevive commented:
`0.10.0` is out. It should be a straightforward update.

armanbilge commented:
so fast!!

alexcardell commented:
Updated

alexcardell committed:
Move KernelSources to KernelSource companion

armanbilge commented on build.sbt:
```diff
@@ -171,6 +174,34 @@ lazy val lambdaCloudFormationCustomResource = crossProject(JSPlatform, JVMPlatfo
+    libraryDependencies ++= Seq(
+      "org.typelevel" %%% "otel4s-core-trace" % otel4sVersion,
+      "org.typelevel" %%% "otel4s-sdk-trace-testkit" % otel4sVersion % Test,
+      "org.typelevel" %%% "otel4s-semconv-experimental" % otel4sVersion % Test,
```
> Experimental (incubating) semantic conventions. Breaking changes expected. Library instrumentation SHOULD NOT depend on this.

@iRevive should we be concerned?

armanbilge commented on build.sbt:
```diff
@@ -179,21 +210,21 @@ lazy val examples = crossProject(JSPlatform, JVMPlatform)
       "org.http4s" %%% "http4s-ember-client" % http4sVersion,
       "org.tpolecat" %%% "natchez-xray" % natchezVersion,
       "org.tpolecat" %%% "natchez-http4s" % "0.6.0",
-      "org.tpolecat" %%% "skunk-core" % "0.6.4"
+      "org.tpolecat" %%% "skunk-core" % "0.6.3"
```
Why is this rolled back?

alexcardell replied:
The answer to this question is lost to time

If there was a reason I've forgotten, but it's probably a mistake

armanbilge commented on lambda-otel4s/shared/src/main/scala/feral/lambda/otel4s/TracedHandler.scala:
```diff
@@ -0,0 +1,53 @@
+object TracedHandler {
+
+  def apply[F[_]: Monad: Tracer, Event, Result](
+      handler: Event => F[Option[Result]]
```
```suggestion
      handler: F[Option[Result]]
```

We don't need to pass the event explicitly, because the user obtains it from the `Invocation`.

alexcardell replied:
Updated

armanbilge commented on lambda-natchez/shared/src/main/scala/feral/lambda/natchez/AwsTags.scala:
```diff
@@ -14,7 +14,7 @@
  */
 
-package feral.lambda
+package feral.lambda.natchez
```
Can we add Scalafix migrations for these renames?

alexcardell replied:
Added

alexcardell replied:
Okay so I've added inputs and outputs but what's the common practice for adding the new v0.3.0 dependency, to run those tests? Do we require a 0.2->0.3 subproject and a 0.3->0.4 subproject?

alexcardell replied:
I see the examples in cats have subprojects per version, I will work on that later

armanbilge replied:
Instead of more subprojects, I wonder if we can just delete / replace the old migrations with the new ones since we are doing a version bump? I would prefer not to maintain them forever esp. if Scalafix/Scalameta may issue deprecations in the future 🤔 

These migrations are already configured in Scala Steward with a hard-coded version.

https://github.com/scala-steward-org/scala-steward/blob/0765eec5afc158c2e30f5740b4aa23fc7ae0b7dd/modules/core/src/main/resources/scalafix-migrations.conf#L250-L255

alexcardell replied:
Removed

armanbilge commented on lambda-otel4s/jvm/src/test/scala/feral/lambda/otel4s/TracedHandlerSuite.scala:
```diff
@@ -0,0 +1,205 @@
+import java.io.ByteArrayOutputStream
+import java.util.concurrent.atomic.AtomicInteger
+
+class TracedHandlerSuite extends CatsEffectSuite {
```
Can we avoid the substantial duplication/splitting of these suite? It looks like you based these on some tests that needed extra machinery because they were testing the platform-specific integration. But I think in your situation it should be possible to test `TracedHandler` using pure code.

alexcardell replied:
Sure thing, makes sense

alexcardell committed:
Remove Event parameter from TracedHandler

alexcardell committed:
Use shared trace handler suite

alexcardell committed:
Add v0.4.0 rewrites for natchez module

alexcardell committed:
Fix formatting

alexcardell committed:
Add scalafix v0.4.0 rewrite input/output

alexcardell committed:
Split 0.3.0 and 0.4.0 rewrite rules into subprojects

alexcardell committed:
Update workflows

alexcardell committed:
Fix scalafix 0.4.0 rewrites test

alexcardell committed:
Remove 0.3.0 migrations

alexcardell committed:
Remove unused examples imports

alexcardell committed:
Rename EventSpanAttributes -> EventAttributeSources

iRevive commented:
There is a compatible pre-release of the http4s-otel4s - https://github.com/http4s/http4s-otel4s-middleware/releases/tag/v0.9.0-rc.2.

alexcardell committed:
Add http4s-otel4s-middleware

alexcardell committed:
Simplify otel attributes file structure

alexcardell committed:
Fix imports

alexcardell committed:
Update example

alexcardell committed:
Fix munit-cats-effect dependency reference

alexcardell committed:
Fix fixture

alexcardell committed:
fix deprecation

alexcardell commented:
It's fixed for 0.13.x

alexcardell commented:
Updated for 0.14.x

alexcardell committed:
Update mergify

jarrodcodes commented:
any chance we could get this merged soon? it works when tested locally.

iRevive commented:
@armanbilge can we move forward with this one?

jarrodcodes commented:
Another bump on this, would be very useful

alexcardell committed:
Bump

alexcardell committed:
fmt

alexcardell committed:
Fix lambda plugin scala version

alexcardell commented:
bumped

apologies for the force push, rebasing was a pain

alexcardell committed:
Add comment explaining semanticdb

iRevive commented on lambda-otel4s/shared/src/main/scala/feral/lambda/otel4s/TracedHandler.scala:
```diff
@@ -0,0 +1,53 @@
+
+object TracedHandler {
+
+  def apply[F[_]: Monad: Tracer, Event, Result](
```
We should _eventually_ add an overload that takes `TracerProvider`, e.g.:
```scala
def apply[F[_]: Monad: TracerProvider, Event, Result](...): F[Option[Result]] = 
  TracerProvider[F]
    .tracer("feral.lambda").withVersion(BuildInfo.version).get
    .flatMap { implicit T: Tracer[F] =>
      ... // rest of the code
    }
```

alexcardell replied:
I am happy to swap it over if that's the preferred dependency, I think at the time I wrote this some libraries were accepting Tracer

I know we have some apps that set both up implicitly for that reason, and an overload may cause a problem unless specifically named

iRevive replied:
Would you mind making this change in this branch? 😅 

OTeL now explicitly asks libraries to own the namespace: https://opentelemetry.io/docs/concepts/instrumentation-scope/

> For libraries and frameworks: use the library’s fully qualified name and version as the scope. If you are writing an instrumentation library for a library that has no built-in OpenTelemetry support, use the name and version of the instrumentation library itself.

I think we can replace `Tracer` with `TracerProvider` instead of offering both.

iRevive commented:
@armanbilge looks good to me, would you mind taking a look too?

alexcardell committed:
more semantic conventions

sources:
https://opentelemetry.io/docs/specs/semconv/faas/faas-spans/
https://opentelemetry.io/docs/specs/semconv/faas/aws-lambda/#sqs

alexcardell committed:
Fix unused import

alexcardell committed:
bump otel4s

alexcardell commented:
@iRevive @armanbilge I've updated to 1.0.0-RC1

alexcardell committed:
Bump to 1.0.0

jarrodcodes commented:
Could we get this merged please? This branch, when compiled and published locally, works just fine. And the author has had to keep this PR synced for over 2 years. This seems like the right time.
